application.cpp

```diff
@@ -445,14 +462,22 @@ void application::exec() {
       (void)work;
       bool more = true;
       while( more || io_serv->run_one() ) {
-         while( io_serv->poll_one() ) {}
-         // execute the highest priority item
-         more = pri_queue.execute_highest();
+         if (my->_is_quiting)
+            break;
+         try {
+            while( io_serv->poll_one() ) {}
+            // execute the highest priority item
+            more = pri_queue.execute_highest();
+         } catch(...) {
```

Comment:
I don't think we should silently swallow the exception. Would be nice to log the exception. `std::exception` with `what()` and `...` with `unknown`.

Reply:
@heifner I was not sure what logging option to use. Any suggestion?

Reply:
`cerr` I think is your only option in appbase. Also should log in your catchs of shutdown.

Reply:
cool, will do!

Reply:
Done.

Reply:
`run_one()` can throw just like the others.

I'm not sure if swallowing the exception like this is the right approach. It's not clear if we may swallow an important message or stomp on an error code (see nodeos' main.cpp for example). For one, an exception caught here makes nodeos return SUCCESS, which looks wrong.

Did we consider making the lifetime of `application` more straightforward so dtors and unwinding "just works" instead of dealing with this complication manually? For example, making the user of `application`,
```c++
int main() {
   appbase::application app;
   app.initialize();
   app.exec();
   return 0;
}
```
Then have ctor & dtor of `application` sets the static `appbase::app()` accessor.

Reply:
hm, I guess the above still has the problem where exceptions don't cause plugin_shutdown() to be called. Now I'm not sure what a plugin ought to expect in that regard.

Reply:
oh hm I guess it's fine.. since application dtor could run the plugin_shutdown()s in that case

Reply:
That would require changing all existing uses of appbase, which assumes a singleton accessed via `app()`.

Reply:
> run_one() can throw just like the others.

If run_one() job is just to enqueue the handler onto the priority queue, the odds of it throwing are pretty low. But if you like I can add another try/catch block including the whole loop.

Reply:
I believe the changes would be limited to something like a main.cpp. It's not like the entirety of nodeos would need to be changed. It would seem to go a long way simplifying this code to make it easier to reason about and easier to read.

Reply:
> I believe the changes would be limited to something like a main.cpp. It's not like the entirety of nodeos would need to be changed. It would seem to go a long way simplifying this code to make it easier to reason about and easier to read.

Instances of `app()` usage outside appbase:

```
./programs/keosd/main.cpp:   auto config_path = app().get_logging_conf();
./programs/keosd/main.cpp:   fc::log_config::initialize_appenders(app().get_io_service());
./programs/keosd/main.cpp:   auto config_path = app().get_logging_conf();
./programs/keosd/main.cpp:   fc::log_config::initialize_appenders(app().get_io_service());
./programs/keosd/main.cpp:   app().set_sighup_callback(logging_conf_handler);
./programs/keosd/main.cpp:      app().set_version_string(eosio::version::version_client());
./programs/keosd/main.cpp:      app().set_full_version_string(eosio::version::version_full());
./programs/keosd/main.cpp:      app().set_default_data_dir(home / "eosio-wallet");
./programs/keosd/main.cpp:      app().set_default_config_dir(home / "eosio-wallet");
./programs/keosd/main.cpp:         .server_header = keosd::config::key_store_executable_name + "/" + app().version_string()
./programs/keosd/main.cpp:      app().register_plugin<wallet_api_plugin>();
./programs/keosd/main.cpp:      if(!app().initialize<wallet_plugin, wallet_api_plugin, http_plugin>(argc, argv)) {
./programs/keosd/main.cpp:         const auto &opts = app().get_options();
./programs/keosd/main.cpp:      auto& http = app().get_plugin<http_plugin>();
./programs/keosd/main.cpp:                       [&a=app()](string, string, url_response_callback cb) {
./programs/keosd/main.cpp:      app().startup();
./programs/keosd/main.cpp:      app().exec();
./programs/nodeos/test/test_nodeos.cpp:   auto config_path = app().get_logging_conf();
./programs/nodeos/test/test_nodeos.cpp:   fc::log_config::initialize_appenders( app().get_io_service() );
./programs/nodeos/test/test_nodeos.cpp:   auto config_path = app().get_logging_conf();
./programs/nodeos/test/test_nodeos.cpp:   fc::log_config::initialize_appenders( app().get_io_service() );
./programs/nodeos/test/test_nodeos.cpp:   app().set_sighup_callback(logging_conf_handler);
./programs/nodeos/test/test_nodeos.cpp:      app().set_version(htonl(short_hash));
./programs/nodeos/test/test_nodeos.cpp:      app().set_version_string(eosio::version::version_client());
./programs/nodeos/test/test_nodeos.cpp:      app().set_full_version_string(eosio::version::version_full());
./programs/nodeos/test/test_nodeos.cpp:      app().set_default_data_dir(root / "eosio" / nodeos::config::node_executable_name / "data" );
./programs/nodeos/test/test_nodeos.cpp:      app().set_default_config_dir(root / "eosio" / nodeos::config::node_executable_name / "config" );
./programs/nodeos/test/test_nodeos.cpp:         .server_header = nodeos::config::node_executable_name + "/" + app().version_string()
./programs/nodeos/test/test_nodeos.cpp:      if(!app().initialize<chain_plugin, net_plugin, producer_plugin, resource_monitor_plugin>(argc, argv)) {
./programs/nodeos/test/test_nodeos.cpp:         const auto& opts = app().get_options();
./programs/nodeos/test/test_nodeos.cpp:      if (auto resmon_plugin = app().find_plugin<resource_monitor_plugin>()) {
./programs/nodeos/test/test_nodeos.cpp:         resmon_plugin->monitor_directory(app().data_dir());
./programs/nodeos/test/test_nodeos.cpp:            ("name", nodeos::config::node_executable_name)("ver", app().version_string())
./programs/nodeos/test/test_nodeos.cpp:            ("fv", app().version_string() == app().full_version_string() ? "" : app().full_version_string()) );
./programs/nodeos/test/test_nodeos.cpp:      ilog("${name} using configuration file ${c}", ("name", nodeos::config::node_executable_name)("c", app().full_config_file_path().string()));
./programs/nodeos/test/test_nodeos.cpp:      ilog("${name} data directory is ${d}", ("name", nodeos::config::node_executable_name)("d", app().data_dir().string()));
./programs/nodeos/test/test_nodeos.cpp:      app().startup();
./programs/nodeos/test/test_nodeos.cpp:      app().set_thread_priority_max();
./programs/nodeos/test/test_nodeos.cpp:      app().exec();
./programs/nodeos/main.cpp:   auto config_path = app().get_logging_conf();
./programs/nodeos/main.cpp:   fc::log_config::initialize_appenders( app().get_io_service() );
./programs/nodeos/main.cpp:   auto config_path = app().get_logging_conf();
./programs/nodeos/main.cpp:   fc::log_config::initialize_appenders( app().get_io_service() );
./programs/nodeos/main.cpp:   app().set_sighup_callback(logging_conf_handler);
./programs/nodeos/main.cpp:      app().set_version(htonl(short_hash));
./programs/nodeos/main.cpp:      app().set_version_string(eosio::version::version_client());
./programs/nodeos/main.cpp:      app().set_full_version_string(eosio::version::version_full());
./programs/nodeos/main.cpp:      app().set_default_data_dir(root / "eosio" / nodeos::config::node_executable_name / "data" );
./programs/nodeos/main.cpp:      app().set_default_config_dir(root / "eosio" / nodeos::config::node_executable_name / "config" );
./programs/nodeos/main.cpp:         .server_header = nodeos::config::node_executable_name + "/" + app().version_string()
./programs/nodeos/main.cpp:      if(!app().initialize<chain_plugin, net_plugin, producer_plugin, resource_monitor_plugin>(argc, argv)) {
./programs/nodeos/main.cpp:         const auto& opts = app().get_options();
./programs/nodeos/main.cpp:      if (auto resmon_plugin = app().find_plugin<resource_monitor_plugin>()) {
./programs/nodeos/main.cpp:         resmon_plugin->monitor_directory(app().data_dir());
./programs/nodeos/main.cpp:            ("name", nodeos::config::node_executable_name)("ver", app().version_string())
./programs/nodeos/main.cpp:            ("fv", app().version_string() == app().full_version_string() ? "" : app().full_version_string()) );
./programs/nodeos/main.cpp:      ilog("${name} using configuration file ${c}", ("name", nodeos::config::node_executable_name)("c", app().full_config_file_path().string()));
./programs/nodeos/main.cpp:      ilog("${name} data directory is ${d}", ("name", nodeos::config::node_executable_name)("d", app().data_dir().string()));
./programs/nodeos/main.cpp:      app().startup();
./programs/nodeos/main.cpp:      app().set_thread_priority_max();
./programs/nodeos/main.cpp:      app().exec();
./plugins/wallet_plugin/wallet_manager.cpp:            appbase::app().quit();
./plugins/wallet_plugin/wallet_manager.cpp:   auto timer = std::make_shared<boost::asio::deadline_timer>(appbase::app().get_io_service(), boost::posix_time::seconds(1));
./plugins/wallet_plugin/wallet_plugin.cpp:static auto _wallet_plugin = app().register_plugin<wallet_plugin>();
./plugins/wallet_plugin/wallet_plugin.cpp:            dir = app().data_dir() / dir;
./plugins/http_plugin/http_plugin.cpp:   static auto _http_plugin = app().register_plugin<http_plugin>();
./plugins/http_plugin/http_plugin.cpp:          * Make an internal_url_handler that will run the url_handler on the app() thread and then
./plugins/http_plugin/http_plugin.cpp:               app().post( priority, [next_ptr, conn=std::move(conn), r=std::move(r), b = std::move(b), wrapped_then=std::move(wrapped_then)]() mutable {
./plugins/http_plugin/http_plugin.cpp:                     if( app().is_quiting() ) return; // http_plugin shutting down, do not call callback
./plugins/http_plugin/http_plugin.cpp:      app().register_config_type<https_ecdh_curve_t>();
./plugins/http_plugin/http_plugin.cpp:         tcp::resolver resolver( app().get_io_service());
./plugins/http_plugin/http_plugin.cpp:               sock_path = app().data_dir() / sock_path;
./plugins/http_plugin/http_plugin.cpp:      app().post(appbase::priority::high, [this] ()
./plugins/http_plugin/http_plugin.cpp:            app().quit();
./plugins/producer_plugin/producer_plugin.cpp:static auto _producer_plugin = app().register_plugin<producer_plugin>();
./plugins/producer_plugin/producer_plugin.cpp:      ,_transaction_ack_channel(app().get_channel<compat::channels::transaction_ack>())
./plugins/producer_plugin/producer_plugin.cpp:            app().get_channel<channels::rejected_block>().publish( priority::medium, block );
./plugins/producer_plugin/producer_plugin.cpp:            appbase::app().quit();
./plugins/producer_plugin/producer_plugin.cpp:               app().post( priority::low, [self, future{std::move(future)}, api_trx, next{std::move( next )}, trx{std::move(trx)}, return_failure_traces]() mutable {
./plugins/producer_plugin/producer_plugin.cpp:   : my(new producer_plugin_impl(app().get_io_service())){
./plugins/producer_plugin/producer_plugin.cpp:               app().get_plugin<signature_provider_plugin>().signature_provider_help_text())
./plugins/producer_plugin/producer_plugin.cpp:   my->chain_plug = app().find_plugin<chain_plugin>();
./plugins/producer_plugin/producer_plugin.cpp:            my->_signature_providers[key_id_to_wif_pair.first] = app().get_plugin<signature_provider_plugin>().signature_provider_for_private_key(key_id_to_wif_pair.second);
./plugins/producer_plugin/producer_plugin.cpp:            const auto& [pubkey, provider] = app().get_plugin<signature_provider_plugin>().signature_provider_for_specification(key_spec_pair);
./plugins/producer_plugin/producer_plugin.cpp:         my->_snapshots_dir = app().data_dir() / sd;
./plugins/producer_plugin/producer_plugin.cpp:      if (auto resmon_plugin = app().find_plugin<resource_monitor_plugin>()) {
./plugins/producer_plugin/producer_plugin.cpp:   my->_incoming_block_subscription = app().get_channel<incoming::channels::block>().subscribe(
./plugins/producer_plugin/producer_plugin.cpp:   my->_incoming_transaction_subscription = app().get_channel<incoming::channels::transaction>().subscribe(
./plugins/producer_plugin/producer_plugin.cpp:   my->_incoming_block_sync_provider = app().get_method<incoming::methods::block_sync>().register_provider(
./plugins/producer_plugin/producer_plugin.cpp:   my->_incoming_transaction_async_provider = app().get_method<incoming::methods::transaction_async>().register_provider(
./plugins/producer_plugin/producer_plugin.cpp:   app().post( 0, [me = my](){} ); // keep my pointer alive until queue is drained
./plugins/producer_plugin/producer_plugin.cpp:      app().quit();
./plugins/producer_plugin/producer_plugin.cpp:         if( app().is_quiting() ) // db guard exception above in LOG_AND_DROP could have called app().quit()
./plugins/producer_plugin/producer_plugin.cpp:      _timer.async_wait( app().get_priority_queue().wrap( priority::high,
./plugins/producer_plugin/producer_plugin.cpp:   _timer.async_wait( app().get_priority_queue().wrap( priority::high,
./plugins/producer_plugin/producer_plugin.cpp:      _timer.async_wait( app().get_priority_queue().wrap( priority::high,
./plugins/producer_plugin/test/test_options.cpp:      appbase::app().initialize<chain_plugin, producer_plugin>( argv.size(), (char**) &argv[0] );
./plugins/producer_plugin/test/test_options.cpp:      appbase::app().startup();
./plugins/producer_plugin/test/test_options.cpp:         {appbase::app().find_plugin<producer_plugin>(), appbase::app().find_plugin<chain_plugin>()} );
./plugins/producer_plugin/test/test_options.cpp:      appbase::app().exec();
./plugins/producer_plugin/test/test_options.cpp:   appbase::app().quit();
./plugins/producer_plugin/test/test_trx_full.cpp:         appbase::app().initialize<chain_plugin, producer_plugin>( argv.size(), (char**) &argv[0] );
./plugins/producer_plugin/test/test_trx_full.cpp:         appbase::app().startup();
./plugins/producer_plugin/test/test_trx_full.cpp:               {appbase::app().find_plugin<producer_plugin>(), appbase::app().find_plugin<chain_plugin>()} );
./plugins/producer_plugin/test/test_trx_full.cpp:         appbase::app().exec();
./plugins/producer_plugin/test/test_trx_full.cpp:            appbase::app().get_channel<plugin_interface::compat::channels::transaction_ack>().subscribe(
./plugins/producer_plugin/test/test_trx_full.cpp:         app().post( priority::low, [ptrx, &next_calls, &num_posts, &trace_with_except, &trx_match, &trxs]() {
./plugins/producer_plugin/test/test_trx_full.cpp:            app().get_method<plugin_interface::incoming::methods::transaction_async>()(ptrx,
./plugins/producer_plugin/test/test_trx_full.cpp:            app().post(priority::high, [](){
./plugins/producer_plugin/test/test_trx_full.cpp:               app().find_plugin<producer_plugin>()->get_integrity_hash();
./plugins/producer_plugin/test/test_trx_full.cpp:      appbase::app().quit();
./plugins/signature_provider_plugin/signature_provider_plugin.cpp:   static auto _signature_provider_plugin = app().register_plugin<signature_provider_plugin>();
./plugins/signature_provider_plugin/signature_provider_plugin.cpp:            return app().get_plugin<http_client_plugin>().get_client().post_sync(keosd_url, params, deadline).as<chain::signature_type>();
./plugins/net_api_plugin/net_api_plugin.cpp:static auto _net_api_plugin = app().register_plugin<net_api_plugin>();
./plugins/net_api_plugin/net_api_plugin.cpp:   auto& net_mgr = app().get_plugin<net_plugin>();
./plugins/net_api_plugin/net_api_plugin.cpp:   app().get_plugin<http_plugin>().add_api({
./plugins/net_api_plugin/net_api_plugin.cpp:      const auto& _http_plugin = app().get_plugin<http_plugin>();
./plugins/producer_api_plugin/producer_api_plugin.cpp:static auto _producer_api_plugin = app().register_plugin<producer_api_plugin>();
./plugins/producer_api_plugin/producer_api_plugin.cpp:   auto& producer = app().get_plugin<producer_plugin>();
./plugins/producer_api_plugin/producer_api_plugin.cpp:   auto& http = app().get_plugin<http_plugin>();
./plugins/producer_api_plugin/producer_api_plugin.cpp:   app().get_plugin<http_plugin>().add_api({
./plugins/producer_api_plugin/producer_api_plugin.cpp:      const auto& _http_plugin = app().get_plugin<http_plugin>();
./plugins/trace_api_plugin/trace_api_plugin.cpp:   static auto plugin_reg = app().register_plugin<trace_api_plugin>();
./plugins/trace_api_plugin/trace_api_plugin.cpp:         trace_dir = app().data_dir() / dir_option;
./plugins/trace_api_plugin/trace_api_plugin.cpp:      if (auto resmon_plugin = app().find_plugin<resource_monitor_plugin>())
./plugins/trace_api_plugin/trace_api_plugin.cpp:               auto abi = abi_def_from_file(kv.second, app().data_dir());
./plugins/trace_api_plugin/trace_api_plugin.cpp:      auto& http = app().get_plugin<http_plugin>();
./plugins/trace_api_plugin/trace_api_plugin.cpp:         app().quit();
./plugins/trace_api_plugin/trace_api_plugin.cpp:      auto& chain = app().find_plugin<chain_plugin>()->chain();
./plugins/test_control_api_plugin/test_control_api_plugin.cpp:static auto _test_control_api_plugin = app().register_plugin<test_control_api_plugin>();
./plugins/test_control_api_plugin/test_control_api_plugin.cpp:   my.reset(new test_control_api_plugin_impl(app().get_plugin<chain_plugin>().chain()));
./plugins/test_control_api_plugin/test_control_api_plugin.cpp:   auto rw_api = app().get_plugin<test_control_plugin>().get_read_write_api();
./plugins/test_control_api_plugin/test_control_api_plugin.cpp:   app().get_plugin<http_plugin>().add_api({
./plugins/net_plugin/net_plugin.cpp:   static auto _net_plugin = app().register_plugin<net_plugin>();
./plugins/net_plugin/net_plugin.cpp:         app().quit();
./plugins/net_plugin/net_plugin.cpp:      app().post( priority::medium, [chain_plug = my_impl->chain_plug, c = shared_from_this(),
./plugins/net_plugin/net_plugin.cpp:      app().post( priority::medium, [blkid, weak{std::move(weak)}]() {
./plugins/net_plugin/net_plugin.cpp:      app().post( priority::medium, [num, weak{std::move(weak)}]() {
./plugins/net_plugin/net_plugin.cpp:         app().post( priority::medium, [chain_plug = my_impl->chain_plug, c,
./plugins/net_plugin/net_plugin.cpp:      if( app().is_quiting() ) {
./plugins/net_plugin/net_plugin.cpp:         app().post( priority::medium, [peer_lib, chain_plug = my_impl->chain_plug, weak{std::move(weak)},
./plugins/net_plugin/net_plugin.cpp:      app().post(priority::medium, [ptr{std::move(ptr)}, id, c = shared_from_this()]() mutable {
./plugins/net_plugin/net_plugin.cpp:      /*producer_plugin* pp = app().find_plugin<producer_plugin>();
./plugins/net_plugin/net_plugin.cpp:         my->chain_plug = app().find_plugin<chain_plugin>();
./plugins/net_plugin/net_plugin.cpp:      my->producer_plug = app().find_plugin<producer_plugin>();
./plugins/net_plugin/net_plugin.cpp:      my->incoming_transaction_ack_subscription = app().get_channel<compat::channels::transaction_ack>().subscribe(
./plugins/net_plugin/net_plugin.cpp:      app().post(priority::highest, [my=my, listen_endpoint](){
./plugins/net_plugin/net_plugin.cpp:               app().quit();
./plugins/net_plugin/net_plugin.cpp:         app().post( 0, [me = my](){} ); // keep my pointer alive until queue is drained
./plugins/state_history_plugin/state_history_plugin.cpp:static auto _state_history_plugin = app().register_plugin<state_history_plugin>();
./plugins/state_history_plugin/state_history_plugin.cpp:                   app().post(priority::medium,
./plugins/state_history_plugin/state_history_plugin.cpp:            app().post(priority::medium, [self = this->shared_from_this()]() {
./plugins/state_history_plugin/state_history_plugin.cpp:         app().post(priority::high,
./plugins/state_history_plugin/state_history_plugin.cpp:            unixs::socket             test_socket(app().get_io_service());
./plugins/state_history_plugin/state_history_plugin.cpp:            app().post(priority::high, [self, s]() mutable { self->session_set.insert(std::move(s)); });
./plugins/state_history_plugin/state_history_plugin.cpp:         // Both app().quit() and exception throwing are required. Without app().quit(),
./plugins/state_history_plugin/state_history_plugin.cpp:         appbase::app().quit();
./plugins/state_history_plugin/state_history_plugin.cpp:      my->chain_plug = app().find_plugin<chain_plugin>();
./plugins/state_history_plugin/state_history_plugin.cpp:         state_history_dir = app().data_dir() / dir_option;
./plugins/state_history_plugin/state_history_plugin.cpp:      if (auto resmon_plugin = app().find_plugin<resource_monitor_plugin>())
./plugins/state_history_plugin/state_history_plugin.cpp:            sock_path = app().data_dir() / sock_path;
./plugins/state_history_plugin/state_history_plugin.cpp:      appbase::app().quit();
./plugins/wallet_api_plugin/wallet_api_plugin.cpp:static auto _wallet_api_plugin = app().register_plugin<wallet_api_plugin>();
./plugins/wallet_api_plugin/wallet_api_plugin.cpp:   auto& wallet_mgr = app().get_plugin<wallet_plugin>().get_wallet_manager();
./plugins/wallet_api_plugin/wallet_api_plugin.cpp:   app().get_plugin<http_plugin>().add_api({
./plugins/wallet_api_plugin/wallet_api_plugin.cpp:      const auto& _http_plugin = app().get_plugin<http_plugin>();
./plugins/template_plugin/template_plugin.cpp:   static auto _template_plugin = app().register_plugin<template_plugin>();
./plugins/test_control_plugin/test_control_plugin.cpp:static auto _test_control_plugin = app().register_plugin<test_control_plugin>();
./plugins/test_control_plugin/test_control_plugin.cpp:         app().quit();
./plugins/test_control_plugin/test_control_plugin.cpp:   my.reset(new test_control_plugin_impl(app().get_plugin<chain_plugin>().chain()));
./plugins/chain_api_plugin/chain_api_plugin.cpp:static auto _chain_api_plugin = app().register_plugin<chain_api_plugin>();
./plugins/chain_api_plugin/chain_api_plugin.cpp:   my.reset(new chain_api_plugin_impl(app().get_plugin<chain_plugin>().chain()));
./plugins/chain_api_plugin/chain_api_plugin.cpp:   auto& chain = app().get_plugin<chain_plugin>();
./plugins/chain_api_plugin/chain_api_plugin.cpp:   auto& http = app().get_plugin<http_plugin>();
./plugins/chain_api_plugin/chain_api_plugin.cpp:   auto& _http_plugin = app().get_plugin<http_plugin>();
./plugins/db_size_api_plugin/db_size_api_plugin.cpp:static auto _db_size_api_plugin = app().register_plugin<db_size_api_plugin>();
./plugins/db_size_api_plugin/db_size_api_plugin.cpp:   app().get_plugin<http_plugin>().add_api({
./plugins/db_size_api_plugin/db_size_api_plugin.cpp:   const chainbase::database& db = app().get_plugin<chain_plugin>().chain().db();
./plugins/resource_monitor_plugin/include/eosio/resource_monitor_plugin/file_space_handler.hpp:         appbase::app().quit(); // This will gracefully stop Nodeos
./plugins/resource_monitor_plugin/resource_monitor_plugin.cpp:   static auto _resource_monitor_plugin = app().register_plugin<resource_monitor_plugin>();
./plugins/chain_plugin/trx_retry_db.cpp:   , _transaction_ack_channel(appbase::app().get_channel<chain::plugin_interface::compat::channels::transaction_ack>())
./plugins/chain_plugin/chain_plugin.cpp:   :pre_accepted_block_channel(app().get_channel<channels::pre_accepted_block>())
./plugins/chain_plugin/chain_plugin.cpp:   ,accepted_block_header_channel(app().get_channel<channels::accepted_block_header>())
./plugins/chain_plugin/chain_plugin.cpp:   ,accepted_block_channel(app().get_channel<channels::accepted_block>())
./plugins/chain_plugin/chain_plugin.cpp:   ,irreversible_block_channel(app().get_channel<channels::irreversible_block>())
./plugins/chain_plugin/chain_plugin.cpp:   ,accepted_transaction_channel(app().get_channel<channels::accepted_transaction>())
./plugins/chain_plugin/chain_plugin.cpp:   ,applied_transaction_channel(app().get_channel<channels::applied_transaction>())
./plugins/chain_plugin/chain_plugin.cpp:   ,incoming_block_channel(app().get_channel<incoming::channels::block>())
./plugins/chain_plugin/chain_plugin.cpp:   ,incoming_block_sync_method(app().get_method<incoming::methods::block_sync>())
./plugins/chain_plugin/chain_plugin.cpp:   ,incoming_transaction_async_method(app().get_method<incoming::methods::transaction_async>())
./plugins/chain_plugin/chain_plugin.cpp:   app().register_config_type<eosio::chain::db_read_mode>();
./plugins/chain_plugin/chain_plugin.cpp:   app().register_config_type<eosio::chain::validation_mode>();
./plugins/chain_plugin/chain_plugin.cpp:   app().register_config_type<chainbase::pinnable_mapped_file::map_mode>();
./plugins/chain_plugin/chain_plugin.cpp:   app().register_config_type<eosio::chain::wasm_interface::vm_type>();
./plugins/chain_plugin/chain_plugin.cpp:            my->blocks_dir = app().data_dir() / bld;
./plugins/chain_plugin/chain_plugin.cpp:            my->state_dir = app().data_dir() / sd;
./plugins/chain_plugin/chain_plugin.cpp:            protocol_features_dir = app().config_dir() / pfd;
./plugins/chain_plugin/chain_plugin.cpp:      if (auto resmon_plugin = app().find_plugin<resource_monitor_plugin>()) {
./plugins/chain_plugin/chain_plugin.cpp:      my->get_block_by_number_provider = app().get_method<methods::get_block_by_number>().register_provider(
./plugins/chain_plugin/chain_plugin.cpp:      my->get_block_by_id_provider = app().get_method<methods::get_block_by_id>().register_provider(
./plugins/chain_plugin/chain_plugin.cpp:      my->get_head_block_id_provider = app().get_method<methods::get_head_block_id>().register_provider( [this]() {
./plugins/chain_plugin/chain_plugin.cpp:      my->get_last_irreversible_block_number_provider = app().get_method<methods::get_last_irreversible_block_number>().register_provider(
./plugins/chain_plugin/chain_plugin.cpp:      my->producer_plug = app().find_plugin<producer_plugin>();
./plugins/chain_plugin/chain_plugin.cpp:      auto shutdown = [](){ return app().quit(); };
./plugins/chain_plugin/chain_plugin.cpp:      auto check_shutdown = [](){ return app().is_quiting(); };
./plugins/chain_plugin/chain_plugin.cpp:   if(app().is_quiting())
./plugins/chain_plugin/chain_plugin.cpp:   app().quit();
./plugins/chain_plugin/chain_plugin.cpp:      itoh(static_cast<uint32_t>(app().version())),
./plugins/chain_plugin/chain_plugin.cpp:      app().version_string(),
./plugins/chain_plugin/chain_plugin.cpp:      app().full_version_string(),
./plugins/chain_plugin/chain_plugin.cpp:      app().get_method<incoming::methods::block_sync>()(std::make_shared<signed_block>( std::move( params ) ), std::optional<block_id_type>{});
./plugins/chain_plugin/chain_plugin.cpp:      app().get_method<incoming::methods::transaction_async>()(pretty_input, true, transaction_metadata::trx_type::input, false,
./plugins/chain_plugin/chain_plugin.cpp:      app().get_method<incoming::methods::transaction_async>()(pretty_input, true, transaction_metadata::trx_type::input, false,
./plugins/chain_plugin/chain_plugin.cpp:      app().get_method<incoming::methods::transaction_async>()(ptrx, true, transaction_metadata::trx_type::input, static_cast<bool>(params.return_failure_trace),
./plugins/chain_plugin/chain_plugin.cpp:        app().get_method<incoming::methods::transaction_async>()(pretty_input, false, transaction_metadata::trx_type::dry_run, true,
./plugins/chain_plugin/test/test_trx_retry_db.cpp:      // run app() so that channels::transaction_ack work
./plugins/chain_plugin/test/test_trx_retry_db.cpp:         appbase::app().initialize( argv.size(), (char**) &argv[0] );
./plugins/chain_plugin/test/test_trx_retry_db.cpp:         appbase::app().startup();
./plugins/chain_plugin/test/test_trx_retry_db.cpp:         plugin_promise.set_value(appbase::app().find_plugin<chain_plugin>());
./plugins/chain_plugin/test/test_trx_retry_db.cpp:         appbase::app().exec();
./plugins/chain_plugin/test/test_trx_retry_db.cpp:            appbase::app().get_channel<plugin_interface::compat::channels::transaction_ack>().subscribe(
./plugins/chain_plugin/test/test_trx_retry_db.cpp:      appbase::app().quit();
```

Reply:
> Then have ctor & dtor of `application` sets the static `appbase::app()` accessor

Although `register_plugin<>()` would require some attention too.

But wow.. _is that it_? Then geez maybe it does make sense to completely remove the static accessor. It's not that hard for the `plugin` to be passed the `application` at construction time, and then it can stash an `application&` somewhere. static variables like this are pretty miserable, as we're seeing first hand.

But really what it comes down to is that the shutdown mechanics are already super convoluted and broken, and what we've done is just adding _more_ complexity, whereas using C++ object lifecycle the standard way would be _removing_ complexity. But if others are fine with it and are confident in these changes then I'm fine for now.

Reply:
I really like the idea of getting rid of the singleton. I'm not sure how you can handle third-party plugins without it though. I guess it could be done by having the cmake `EOSIO_ADDITIONAL_PLUGINS` modify a header that main includes that registers the plugins. Probably not too terrible.

Reply:
> whereas using C++ object lifecycle the standard way would be removing complexity.

While it would certainly be cleaner to not use static variables, I don't believe it would remove as much complexity as you suggest. For example we'd still want to iterate through the shutdown and destruction of the plugins, so that if there is an exception in one of them, the others still get shutdown correctly. 

I think most of the current additions are improvements which would be needed regardless.

Reply:
>  I guess it could be done by having the cmake EOSIO_ADDITIONAL_PLUGINS modify a header that main includes that registers the plugins.

That seems maybe even worse to me than the static variable. If we wanted to do that maybe another issue and PR would be appropriate.